typegraph/core/wit/typegraph.wit

```diff
@@ -1,28 +1,101 @@
 
 interface core {
-  record tpe {
-    id: u32
-  }
 
-  union out {
-    tpe,
-    string,
-    s32,
-    u32
-  }
+    type error = string
 
-  integerb: func() -> tpe
+    record typegraph-init-params {
+        name: string,
+    }
 
-  // do a set per wit type, not per typegraph type (similar to gettpe)
-  integermin: func(id: u32, n: s32) -> tpe
+    init-typegraph: func(params: typegraph-init-params) -> result<_, error>
+    finalize-typegraph: func() -> result<string, error>
 
-  structb: func(props: list<tuple<string, tpe>>) -> tpe
+    type type-id = u32
 
-  gettpe: func(id: u32, field: string) -> option<tpe>
+    record type-base {
+        name: option<string>,
+    }
+
+    record type-proxy {
+        name: string,
+    }
+
+    proxyb: func(data: type-proxy) -> result<type-id, error>
+
+    record type-integer {
+        min: option<s32>,
+        max: option<s32>,
+    }
+
+    integerb: func(data: type-integer, base: type-base) -> result<type-id, error>
+    type-as-integer: func(ref: type-id) -> result<tuple<type-id, type-base, type-integer>, error>
+
+    record type-struct {
+        props: list<tuple<string, type-id>>,
+    }
+
+    structb: func(data: type-struct, base: type-base) -> result<type-id, error>
+    type-as-struct: func(id: type-id) -> result<tuple<type-id, type-base, type-struct>, error>
+
+    get-type-repr: func(id: type-id) -> result<string, error>
+
+    record type-func {
+        inp: type-id,
+        out: type-id,
+        mat: materializer-id,
+    }
+    funcb: func(data: type-func) -> result<type-id, error>
+
+    expose: func(fns: list<tuple<string, type-id>>, namespace: list<string>) -> result<_, error>
+    
+    type runtime-id = u32
+    type materializer-id = u32
+
+    register-runtime: func() -> runtime-id
+}
+
+
+interface runtimes {
+    use self.core.{error, runtime-id, materializer-id}
+
+    type idempotency = bool
+
+    variant effect {
+        none,
+        create(idempotency),
+        update(idempotency),
+        delete(idempotency),
+        upsert(idempotency),
+    }
+
+    record base-materializer {
+        runtime: runtime-id,
+        effect: effect,
+    }
+
+    record materializer-deno-func {
```

Comment:
I believe we shall add high-level comments on this file for all types of objects that a developer may want to integrate in his own SDK